poly/promise/promise.go

```diff
@@ -0,0 +1,744 @@
+}
+
+/* This helper function makes sure that a share is "syntactically" valid. In
+ * other words, the index is properly in range and the guardian it was sent to
+ * the correct guardian. VerifyShare handles "semantic" verification.
+ *
+ * Arguments
+ *   i        = the index of the share
+ *   gKeyPair = the key pair of the guardian of the share
+ *
+ * Returns
+ *    Whether the share is valid or invalid.
+ */
+func (p *Promise) validShare(i int, gKeyPair *config.KeyPair) bool {
+	if i < 0 || i > p.n {
+		return false
+	}
+	if !p.guardians[i].Equal(gKeyPair.Public) {
+		return false
+	}
+	return true
+}
+
+
+/* Verify that a share has been properly constructed.
+ *
+ * Arguments
+ *    i        = the index of the share to verify
+ *    gPrikey  = the private key of the guardian of share i
+ *
+ * Return
+ *   whether the decrypted secret properly passes the public polynomial.
+ *
+ * Note
+ *   Make sure that the proper index and key is specified. Otherwise, the
+ *   function will return false because diffieHellmanDecrypt gave the wrong
+ *   result. In short, make sure to verify only shares that are allotted to you.
+ */
+func (p *Promise) VerifyShare(i int, gKeyPair *config.KeyPair) bool {
+	if !p.validShare(i, gKeyPair) {
+		return false
+	}
+	diffieBase := p.shareSuite.Point().Mul(p.pubKey, gKeyPair.Secret)
+	share := p.diffieHellmanDecrypt(p.secrets[i], diffieBase)
+	return p.pubPoly.Check(i, share)
+}
+
+/* Verifies whether a Promise objects has enough signatures to be valid.
+ *
+ * Return
+ *   whether the Promise is now valid and considered trustworthy.
+ *
+ * Technical Notes: The function goes through the list of signatures and checks
+ *                  whether the signature is properly signed. If at least r of
+ *                  these are signed and r is greater than t (the minimum number
+ *                  of shares needed to reconstruct the secret), the promise is
+ *                  considered valid.
+ */
+func (p *Promise) VerifyPromise() bool {
+	validSigs := 0
+	for i := 0; i < p.n; i++ {
+		// Check whether the signature is initialized. Otherwise, bad
+		// things will happen.
+		if p.signatures[i] != nil && p.VerifySignature(p.signatures[i]) {
+			validSigs += 1
+		}
+	}
+	return p.r > p.t && validSigs >= p.r
+}
+
+/* Produce a signature for a given guardian
+ *
+ * Arguments
+ *    i         = the index of the guardian's share
+ *    gKeyPair  = the public/private keypair of the guardian.
+ *
+ * Return
+ *   A PromiseSignature object with the signature.
+ *
+ * Note:
+ *   The signature message will always be of the form:
+ *      Guardian approves PromiseId
+ *
+ *   It is assumed that the guardian has called VerifyShare first and hence
+ *   it is assumed that the input to the function is trusted.
+ */
+func (p *Promise) Sign(i int, gKeyPair *config.KeyPair) *PromiseSignature {
+	set        := anon.Set{gKeyPair.Public}
+	approveMsg := gKeyPair.Public.String() + " approves " + p.id
```

Comment:
I changed the signing to use predefined messages.

My thought behind using the more specific messages was to make it harder for malicious signature reuse. A malicious user who wanted to frame a server could produce a false BlameProof message. It could then get the victim to produce a valid signature for a valid share it has produced. It could then take that signature and attach it to the BlameProof. Hence, it could pretend that the server had unjustly blamed it. Having fine grain messages would prevent a malicious server from reusing signatures in such fashion since the malicious server would need to get the victim to sign a message saying "publicKey blames publicKey" instead of "publicKey approves publicKey".

Do you think this will be a concern?

However, the approach above wouldn't work well if we wanted to do collective signing as I can imagine the message soon growing to an unwieldy size.